### ISSUE TITLE
Relationals do not work with any methods of Expr

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#5031]
```
Consider how none of the following work:

In [25]: Eq(x, y) + Eq(y, z)
Out[25]: x = y + y = z

In [26]: Eq(x, y) * Eq(y, z)
Out[26]: x = y⋅y = z

In [27]: Eq(x, y).diff(x)
Out[27]: 
d        
──(x = y)
dx       

In [30]: Eq((x + y)**2, z).expand()
Out[30]: 
       2    
(x + y)  = z

These are all just methods of Expr (the first is just __add__).  For Eq, anything method of Expr 
should be correct if you just apply Eq(Eq.lhs.method, Eq.rhs.method). Does anyone know of an easy 
way to vectorize all Expr methods for Eq?  Also, we should do similar for the inequalities, but only if 
they are actually correct.
```

Original issue for #5031: http://code.google.com/p/sympy/issues/detail?id=1932
Original author: https://code.google.com/u/asmeurer@gmail.com/
Referenced issues: #5030
Original owner: https://code.google.com/u/asmeurer@gmail.com/

**Comment by oscarbenjamin** [comment]
It's a shame that this issue is languishing. Working with equations is much more natural than working with expressions. It is possible to work with `A - B` rather than `A = B` but it's generally less good.

Consider being able to do something like this:
```julia
In [8]: eq1 = Eq(E**(I*x), cos(x)+I*sin(x))

In [9]: eq1
Out[9]: 
 ⅈ⋅x                    
ℯ    = ⅈ⋅sin(x) + cos(x)

In [11]: eq1.conjugate()
Out[11]: 
________________________
 ⅈ⋅x                    
ℯ    = ⅈ⋅sin(x) + cos(x)

In [12]: (eq1 + eq1.conjugate()) / 2
Out[12]: 
________________________                             
 ⅈ⋅x                       ⎛ ⅈ⋅x                    ⎞
ℯ    = ⅈ⋅sin(x) + cos(x) + ⎝ℯ    = ⅈ⋅sin(x) + cos(x)⎠
─────────────────────────────────────────────────────
                          2  
```

I'm hoping you can imagine what I'd like to see there.

If you're working with an expression as `A - B` as an implicit representation of `A = B` then how do you e.g. take logs of both sides of the equation to get to `log(A) - log(B)`? It's definitely more natural to be able to write `log(eq)` than work with the two sides independently.

The idea stated above that equalities are booleans is incorrect. Equalities are statements and it is much better to work with statements than expressions.

I am happy to write a patch to change this if it is likely to be accepted.

**Comment by hlovatt** [comment]
This would be a great feature. I get round this limitation using a list of two items to represent the lhs and rhs and then use list comprehension to operate on both sides. EG:
```python
e0 = [sum(side) for side in zip([x, y], [y, z])]  # Original example `Eq(x, y) + Eq(y, z)`.
print(f'e0 = {e0}')  # [x + y, y + z]
e1 = [side.diff(x) for side in [x, y]]  # Original example `Eq(x, y).diff(x)`.
print(f'e1 = {e1}')  # [1, 0]
```

**Comment by smichr** [comment]
> equalities are booleans is incorrect

Not in SymPy. `Eq` is an unevaluated Boolean that evalulates to True or False when it is known whether the two sides are the same or now. That being said, however, Ronan also pointed out: "it doesn't really make sense to add them" so we might say that *in context* we can figure out when to treat them differently. 

I think simply leaving them alone as Booleans and taking up #4986 would be the better way to go. Someone already offered a working `Eqn` class that could get the process started.

**Comment by hlovatt** [comment]
If:

1. `x == y` was equivalent to `Eq(x, y)`, i.e. it created an `Eq`.
2. `Eq` had operations so that `(x == y) + (y == z)`, `(x == y).diff(x)`, etc. worked symbolically.
3. Similarly, `Ne`, `Lt`, `Gt`, `Le`, and `Ge` where created by `!=`, `<`, `>`, `<=`, and `>=` respectively and had operations, i.e. they are all types of equations.

Then `N(eq)` could be used to evaluate an equation to a boolean.

Something similar works well in Mathematica.

**Comment by smichr** [comment]
> 1. `x == y` was equivalent to `Eq(x, y)`, i.e. it created an `Eq`.

Hmm. Then all the if `x == y` in the code would have to be changed to `N(x == y)` or perhaps `bool(x==y)`. Others will know if that is doable. I like the idea.

Also, here is an update to the `do` function that can be used in the meantime:
```python
def do(self, e, i=None):
    """Return a new Eq using function given or a model
    model expression in which a variable represents each
    side of the expression.

    Examples
    ========

    >>> from sympy import Eq
    >>> from sympy.abc import i, x, y, z
    >>> eq = Eq(x, y)

    When the argument passed is an expression with one
    free symbol that symbol is used to indicate a "side"
    in the Eq and an Eq will be returned with the sides
    from self replaced in that expression. For example, to
    add 2 to both sides:

    >>> eq.do(i + 2)
    Eq(x + 2, y + 2)

    To add x to both sides:

    >>> eq.do(i + x)
    Eq(2*x, x + y)

    In the preceding it was actually ambiguous whether x or i
    was to be added but the rule is that any symbol that are
    already in the expression are not to be interpreted as the
    dummy variable. If we try to add z to each side, however, an 
    error is raised because now it is unclear whether i or z is being
    added:

    >>> eq.do(i + z)
    Traceback (most recent call last):
    ...
    ValueError: not sure what symbol is being used to represent a side

    The ambiguity must be resolved by indicating with another parameter 
    which is the dummy variable representing a side:

    >>> eq.do(i + z, i)
    Eq(x + z, y + z)

    Alternatively, if only one Dummy symbol appears in the expression then
    it will be automatically used to represent a side of the Eq.

    >>> eq.do(2*Dummy() + z)
    Eq(2*x + z, 2*y + z)

    It is also possible to do Eq/Eq operations:
    >>> eq.do(i + Eq(x, 2))
    Eq(2*x, y + 2)
    """
    if isinstance(e, (FunctionClass, Lambda, type(lambda:1))):
        return self.applyfunc(e)
    e = S(e)
    if len(e.args) == 2:
        a, b = e.args
        if isinstance(a, Equality)  or isinstance(b, Equality):
            if isinstance(b, Symbol):
                return self.func(e.func(a.lhs, self.lhs), e.func(a.rhs, self.rhs))
            if isinstance(a, Symbol):
                return self.func(e.func(self.lhs, b.lhs), e.func(self.rhs, b.rhs))
            raise ValueError('expecting 1 arg to be a symbol')
    i = (set([i]) if i else e.free_symbols) - self.free_symbols
    if len(i) != 1:
        d = [i for i in i if isinstance(i, Dummy)]
        if len(d) != 1:
            raise ValueError('not sure what symbol is being used to represent a side')
        i = set(d)
    i = i.pop()
    f = lambda side: e.subs(i, side)
    return self.func(*[f(side) for side in self.args])

from sympy.core.relational import Equality
Equality.do = do
```

**Comment by hlovatt** [comment]
@smichr I like the idea of `do`. Unfortunately having trouble with `Eq(x, y).do(i.diff(x), i)`. It gives `True`, was expecting `Eq(1, 0)`. Also needs the explicit `i` for some reason. I'm probably doing something wrong!

**Comment by smichr** [comment]
Thinking more about letting `==` give `Eq`. I don't think that can happen at this point since that would be a major regression for SymPy.

2 issue with `do` problem: 
1. `i.diff(x)` evaluates to 0 so you just said `Eq(x,y).do(0)`; when you told it what `i` was it didn't help -- an error should be raised in this case -- because then it replaced each side with `S(0).subs(i, side)` and produced `Eq(0,0)`.
2. It might be good to have evaluate=False in `do` so you can always see the result like `Eq(1, 0)` otherwise it will evaluate to False.

The way to do what you intended is to do `Eq(x, y).do(lambda i: i.diff(x))`.

I updated the do function in the above comment.

**Comment by oscarbenjamin** [comment]
> Thinking more about letting `==` give `Eq`. I don't think that can happen at this point since that would be a major regression for SymPy.

It's often said that == is for structural equality but that's not true since `S(2) == S(2.0)`. The truth is that == is needed for dicts and sets. It is designed to match what Python would do with the corresponding unsimpified objects. That is used heavily throughout sympy and also by users of sympy so I don't think it can be changed. The other operations like < are also used by things like `if a < 2` and in simplification and by sort etc:
```
In [2]: sorted([1, sqrt(2), 3])
Out[2]: [1, √2, 3]
```

I think in a lot of these contexts bool is implicitly called though so I wonder if it would be possible to make so that these relations don't evalute to bool until bool is called on them e.g.:
```
In [10]: S(1) < S(2)
Out[10]: 1 < 2

In [11]: bool(S(1) < S(2))
Out[11]: True
```
Of course this would be fragile since
```
In [15]: 1 < 2
Out[15]: True
```

Maybe a wrapper class could be used as an API for creating relations of each type that don't simplify to bool:
```
In [16]: Declare(1) < 2
Out[16]: 1 < 2

In [16]: Declare(1) == 2
Out[16]: 1 = 2
```

**Comment by smichr** [comment]
> wrapper class could be used

We can already avoid evaluation with `evaluate=False`.

**Comment by oscarbenjamin** [comment]
> > wrapper class could be used
> 
> We can already avoid evaluation with `evaluate=False`.

I want something that will never evaluate unlike this:
```
In [35]: e = Ge(x, 1, evaluate=False)                                                                    

In [36]: e                                                                                               
Out[36]: x ≥ 1

In [37]: e.subs(x, 2)                                                                                    
Out[37]: True

In [38]: e.subs(x, 2, evaluate=False)                                                                    
Out[38]: True
```
I want to be able to do things like `2*e`, `simplify(e)` etc without the relation ever turning into a bool.

**Comment by smichr** [comment]
> without the relation ever turning into a bool.

oh. That's what `do` does...until we have something better. What happens when you make an UnevaluateExpr of a Relational? Can you do so and does it respond to simplification?

**Comment by oscarbenjamin** [comment]
UnevaluateExpr doesn't stop evaluation:
```julia
In [4]: UnevaluatedExpr(Ge(x, 1))
Out[4]: x ≥ 1

In [5]: e = UnevaluatedExpr(Ge(x, 1))

In [6]: e
Out[6]: x ≥ 1

In [7]: e.subs(x, 2)
Out[7]: True
```

Also the arithmetic operations give nonsensical results:
```julia
In [8]: 2*e
Out[8]: 2⋅x ≥ 1

In [9]: e + 1
Out[9]: 1 + x ≥ 1
```
It may look like those operations only apply to the left hand side but that's just because it hasn't printed with brackets really it is an Add/Mul with the inequality as arg:
```julia
In [11]: srepr(2*e)
Out[11]: "Mul(Integer(2), UnevaluatedExpr(GreaterThan(Symbol('x'), Integer(1))))"
```

Simplify still turns them into bools:
```julia
In [13]: UnevaluatedExpr(Ge(2, 1, evaluate=False))
Out[13]: 2 ≥ 1

In [14]: simplify(_)
Out[14]: True
```
I can see the benefit of simplifying the inequality to a bool but I also want a way to work with equations/inequalities. I see now that we need distinct classes to represent these and a convenient API for creating them.

**Comment by czgdp1807** [comment]
I read the whole thread and found operations on equality quite interesting. I have an idea on how this thing can be done. I am just getting familiar with the issue, so may be I can miss some details. 
One option to achieve this is to define operations like `add`, `subtract`, `divide` on the objects of `Eq` by defining magic methods in the associated class for `Eq`. A rough example is presented below.
```
>>> from sympy import *
>>> from sympy.abc import x, y, z
>>> eqn = Eq(x, y)
>>> eqn + 2
Eq(x + 2, y + 2)
>>> eqn/3
Eq(x/3, y/3)
>>> eqn.diff(x)
Eq(x, 0)
```
This is just my approach, since, it's quite intuitive to me while doing mathematics.

**Comment by oscarbenjamin** [comment]
I think what we need is a new equation class i.e. #4986.

With that we can then make Relational not subclass Expr and then these operations with Eq will be a TypeError.

**Comment by czgdp1807** [comment]
@oscarbenjamin The [comment](https://github.com/sympy/sympy/issues/4986#issuecomment-219270679) can solve the problem. However, as @asmeurer said that, the above operations should be done `Eq` rather than defining a new class. Though, please explain how the defining methods, like, `__add__` will effect the current `Eq`?

**Comment by smichr** [comment]
> should be done with `Eq` rather than defining a new class

A conceptual shift must happen. The gut feeling is that Eq should "just work". But the way it is defined, it is a Boolean and it doesn't make sense to do mathematical operations with Booleans like those that are being suggested. And to make it behave differently than it currently does -- e.g. to make it unevaluated by default -- would create a lot of undesirable behavior. For example, no Piecewise would evaluate until you actually gave it the `doit` cocmmand to cause any `Eq` in the conditions to evaluate.

**Comment by czgdp1807** [comment]
@smichr I understood your point. Infact, @oscarbenjamin 's comment made me realise the side effects of defining operations of equality in `Eq`, as `Relational` also inherits it if I am not wrong.
I am also in favor of making a new class, like, `Eqn`. Let's see, what @asmeurer has to say over this point.

**Comment by oscarbenjamin** [comment]
> Piecewise would evaluate until you actually gave it the `doit` command to cause any `Eq` in the conditions to evaluate.

Actually this doesn't seem a problem to me. Piecewise.eval can process the conditions and call bool on them. If `Relation.__bool__` was cached it might even be more efficient in some cases since some Relationals might never need to be evaluated.

I wonder more generally if Relationals could be made to defer evaluation until `__bool__`...

**Comment by czgdp1807** [comment]
@oscarbenjamin One more thing to consider which I thought about. Suppose we use `Eq` both for equality operations and boolean values, then I think it will be against the notion of the definition of a class, which according to me is, a group of objects having similar properties. In mathematics, saying, `a = b` holds a completely different set of properties as compared to Eq(a, b) which can be evaluated to bool by using, probably `subs` which have a different set of properties. I feel defining a new class is the best option to implement equality operations in sympy and it will be manageable too.

**Comment by czgdp1807** [comment]
@asmeurer @smichr @oscarbenjamin I am quite interested to implement this. Please provide your updates what should be done, like, create a new class `Eqn` or make modifications to `Eq`?

**Comment by asmeurer** [comment]
I haven't really thought about this in a while. It may be best to play around with some of the ideas here and see how it affects the existing tests.

**Comment by czgdp1807** [comment]
@asmeurer Thanks for the response. I will make a PR soon on this issue.

**Comment by asmeurer** [comment]
Keeping Eq unevaluated until bool() is called on it could be a clean way to get https://github.com/sympy/sympy/issues/4986 without actually having separate classes. Outside of bool(), it would be inferred from context whether the equation is supposed to be symbolic or boolean. For instance, if you use it in an algebraic expression, it is symbolic, but if it is in a boolean expression (like an And or as the condition of a Piecewise), it is boolean. 

But again I haven't thought about it in a while so there may be some reason why it's not a good idea to keep a single class, aside from the obvious "type confusion" concerns. I guess one problem is that any boolean expression that could contain Eq would be responsible for calling doit() on it if it wants the (current) autoevaluate behavior.

**Comment by oscarbenjamin** [comment]
I think that before creating a new class we should try out having Eq not evaluate to see what happens. A PR with that could explore whether it actually works.

**Comment by curvi** [comment]
People here have highlighted some very important aspects, of which to name a few are

- having an equation would be definately worth it!
- x-y has mathematical flaws (like application of a function to both sides and substitution)
- equations are not booleans
- intuition about what `Eq()` does is quite missleading

In that sense I would ask you to stick with `Eq()` and not open another construction site. As far as I can see the boolean approach can still be there in `equation()` and `Eq()` for a mathematical equation. I just wanted to highlight, that a boolean evaluation is already implemented. And I think, that the fear of breaking something should not be the deciding factor for design decisions.
Thanks for the effort  @czgdp1807 @asmerer I'm glad to see development being active and on a brave way!

**Comment by sylee957** [comment]
I see there was an attempt to override algebraic operations for relationals in #8023
But personally, I think that adding separate methods like `add_sides`, `subtract_sides` or such can make more sense than, having algebraic operations overrided for equations, as I see it from Mathematica.

https://reference.wolfram.com/language/ref/AddSides.html

Also, I find that things like multiplying or dividing have some more things to consider, like dividing by zero.
Multiplying both sides by zero can still make sense in equation, but it may have to be considered special because it makes it impossible to restore the original expression from the evaluated one.

My thoughts are that, there can be two types of algebraic transformations for relationals.
First one can be restricting the output to be equivalent as input. In this case, multiplying both sides with zero should be disallowed. It should be used for the places where both equations should have same solutions.
Second one can be allowing the input to be sufficient condition for the output, even if it breaks the system. In this context, multiplying both sides with zeros can make sense, or even things like adding an equation with inequality can be possible.

I have just worked on implementing some algebraic transformations for a while, following up
https://en.wikipedia.org/wiki/Equation
https://en.wikipedia.org/wiki/Inequality_(mathematics)
and trying Mathematica's implementations.
I may open up a draft PR soon.

**Comment by czgdp1807** [comment]
@sylee957 I read the thread at #8023 and found that people are supporting special methods rather than overriding magic methods for equation manipulations. So, I think we should go with your plan.
Please tag me in your PR so that I can see how the implementation looks like. Waiting for the update.

**Comment by oscarbenjamin** [comment]
I don't see why all operations on an equation have to be invertible. I think it's fine to say that if `x == y` then `x**2 = y**2` etc. That's a standard algebraic manipulation that is often used.

**Comment by asmeurer** [comment]
The problem would be going the other way. But sympy already doesn't simplify `sqrt(x**2)`. Any function that doesn't apply over equations would have issues in normal expression contexts as well.

OTOH it might be wise to not have inequalities be so easy to manipulate, since they are only valid for increasing functions.

**Comment by smichr** [comment]
> inferred from context whether the equation is supposed to be symbolic or boolean

I think I already have this detection working and an error is raise when you try to mix Expr and Bool, so `Eq(x + y, True)` will raise an error because `x + y` is not Boolean.

**Comment by smichr** [comment]
> In this case, multiplying both sides with zero should be disallowed

Or it could create `Piecewise((Eq(zero*y, zero*z), Ne(zero, 0)), (Undefined, True))`

**Comment by czgdp1807** [comment]
> I may open up a draft PR soon.

Any updates on this? @sylee957 Are you working on this?

**Comment by oscarbenjamin** [comment]
There are two separate things that have been discussed in this issue:

1. Relational currently does nonsensical things when used in e.g. arithmetic operations.

2. It would be good to have a SymPy "Equation" class that behaves in a different way and can be manipulated with arithmetic operations.

The first of these problems is due to `Relational` being a subclass of `Expr` which I intend to fix in #18053.

The second issue should be resolved by the introduction of a new class. That is described in #4986 which also includes an implementation: https://github.com/sympy/sympy/issues/4986#issuecomment-219270679

I propose that this can be closed if #18053 is merged and the issue for adding a new `Equation` class can be #4986 which can remain open.
